lib/Sema/TypeCheckGeneric.cpp

```diff
@@ -1274,37 +1280,36 @@ RequirementCheckResult TypeChecker::checkGenericArguments(
     case RequirementKind::Superclass:
       // Superclass requirements.
       if (!isSubclassOf(firstType, secondType, dc)) {
-        if (loc.isValid()) {
-          // FIXME: Poor source-location information.
-          diagnose(loc, diag::type_does_not_inherit, owner, firstType,
-                   secondType);
-
-          diagnose(noteLoc, diag::type_does_not_inherit_requirement,
-                   rawFirstType, rawSecondType,
-                   genericSig->gatherGenericParamBindingsText(
-                       {rawFirstType, rawSecondType}, substitutions));
-        }
-
-        return RequirementCheckResult::Failure;
+        diagnostic = diag::type_does_not_inherit;
+        diagnosticNote = diag::type_does_not_inherit_requirement;
+        requirementFailure = true;
       }
-      continue;
+      break;
 
     case RequirementKind::SameType:
       if (!firstType->isEqual(secondType)) {
-        if (loc.isValid()) {
-          // FIXME: Better location info for both diagnostics.
-          diagnose(loc, diag::types_not_equal, owner, firstType, secondType);
-
-          diagnose(noteLoc, diag::types_not_equal_requirement, rawFirstType,
-                   rawSecondType,
-                   genericSig->gatherGenericParamBindingsText(
-                       {rawFirstType, rawSecondType}, substitutions));
-        }
-
-        return RequirementCheckResult::Failure;
+        diagnostic = diag::types_not_equal;
+        diagnosticNote = diag::types_not_equal_requirement;
+        requirementFailure = true;
       }
+      break;
+    }
+
+    if (!requirementFailure)
       continue;
+
+    if (listener)
+      listener->unsatisfiedRequirement(rawReq, firstType, secondType);
```

Comment:
The implied contract for `unsatisfiedRequirement` is a bit odd... it might diagnose the problem, or it might not, and `checkGenericArguments` will (separately) decide based on the validity of `loc` whether it should diagnose. Could we have `unsatisfiedRequirement` return `true` to indicate when it diagnosed the requirement, in which case `checkGenericArguments` won't emit a diagnostic? Perhaps then we should call it `diagnoseUnsatisfiedRequirement`.